EIPS/eip-6066.md

```diff
@@ -0,0 +1,124 @@
+---
+eip: 6066
+title: Signature Validation Method for NFTs
+description: A way to verify signatures when the signing entity is an EIP-721 or EIP-1155 NFT
+author: Jack Boyuan Xu (@boyuanx)
+discussions-to: https://ethereum-magicians.org/t/eip-6066-signature-validation-method-for-nfts/
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-11-29
+requires: 721, 1155
+---
+
+## Abstract
+
+While **E**xternally **O**wned **A**ccounts can validate signed messages with `ecrecover()` and smart contracts can validate signatures using specifications outlined in [EIP-1271](./eip-1271.md), currently there is no standard method to create or validate signatures made by NFTs. We propose a standard way for anyone to validate whether a signature made by an NFT is valid. This is possible via a modified signature validation function originally found in [EIP-1271](./eip-1271.md): `isValidSignature(tokenId, hash)`.
+
+## Motivation
+
+With billions of ETH in trading volume, the **N**on-**F**ungible **T**oken standard has exploded into tremendous popularity in recent years. Despite the far-reaching implications of having unique tokenized items on-chain, NFTs have mainly been used to represent artwork in the form of avatars or profile pictures. While this is certainly not a trivial use case for the [EIP-721](./eip-721.md) & [EIP-1155](./eip-1155.md) token standards, we reckon more can be done to aid the community in discovering alternative uses for NFTs.
+
+## Specification
+
+_The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119._
+
+```
+pragma solidity ^0.8.0;
+
+interface IEIP6066 {
+    /**
+     * @dev Should return whether the signature provided is valid for the provided tokenId and hash
+     * @param tokenId   Token ID of the signing NFT
+     * @param hash      Hash of the data to be signed
+     *
+     * MUST return the bytes4 magic value 0x1e6395e6 when function passes.
+     * MUST NOT modify state (using STATICCALL for solc < 0.5, view modifier for solc > 0.5)
+     * MUST allow external calls
+     *
+     */
+    function isValidSignature(
+        uint256 tokenId,
+        bytes32 hash
```

Comment:
I am in full support for this EIP which extends EIP-1271, this is well needed! E.g. This EIP will be useful for [EIP-5453](https://eips.ethereum.org/EIPS/eip-5453) endorsement which. Feel free to mention use cases like EIP-5453 in your Motivation to add the weights of necessity of this EIP.

Also, I like to propose considering adding a extension field for future-proof (https://eips.ethereum.org/EIPS/eip-5750)
```suggestion
        bytes32 hash,
        bytes calldata extraData,
```

For example, this will allow future extending EIPs, e.g.
1.  add "only consider valid when user's EIP-1155 balance of tokenId > miniumBalance
2. only consider valid if user also hold slots of `slotId` 

The EIP-5750 documented the rationale why adding an extension is beneficial. Had EIP-1271 has this extending field, they will not have the disadvantage that this EIP intends to solve and improve.

In short, this is EIP a great idea.

Reply:
Thank you for the support and feedback, I have incorporated EIP-5750 into this EIP.